neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt

```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
+                }
+    }
+
+    override fun renewedSubscriptionToPlanSuccessfully(customerId: String,
+                                                       sku: String,
+                                                       subscriptionPaymentInfo: SubscriptionPaymentInfo): Either<StoreError, Plan> = writeTransaction {
         IO {
             Either.monad<StoreError>().binding {
-                val product = get(Product withSku sku).bind()
-                val plan = get(Plan withId sku).bind()
+                val product = get(Product withSku sku)
+                        .bind()
+                val plan = get(Plan withId sku)
+                        .bind()
+
+                /* For successful purchases it should never be the case that the 'charge-id'
+                   is not set. */
+                val dummyChargeId = UUID.randomUUID().toString()
+                if (subscriptionPaymentInfo.chargeId == null) {
+                    logger.error("Got no 'charge-id' with successful renewal of subscription " +
+                            "${subscriptionPaymentInfo.id} to product ${sku} for customer ${customerId} " +
+                            "falling back to using the dummy value ${dummyChargeId}")
+                }
+                val chargeId = subscriptionPaymentInfo.chargeId ?:
+                    dummyChargeId
                 val purchaseRecord = PurchaseRecord(
                         id = chargeId,
                         product = product,
                         timestamp = Instant.now().toEpochMilli(),
-                        properties = mapOf("invoiceId" to invoiceId)
+                        properties = mapOf(
+                                "invoiceId" to (subscriptionPaymentInfo.invoiceId),
+                                "paymentStatus" to subscriptionPaymentInfo.status.name)
                 )
 
                 /* Will exit if an existing purchase record matches on 'invoiceId'. */
                 createPurchaseRecord(customerId, purchaseRecord)
                         .bind()
 
-                // FIXME Moving customer to new segments should be done only based on productClass.
-                /* Offer products to the newly signed up subscriber. */
-                product.segmentIds.forEach { segmentId ->
-                    assignCustomerToSegment(
-                            customerId = customerId,
-                            segmentId = segmentId,
-                            transaction = transaction)
-                            .bind()
-                }
-                logger.info("Customer $customerId completed payment of invoice $invoiceId for subscription to plan ${plan.id}")
-
+                logger.info(
+                        "Customer $customerId completed payment of invoice ${subscriptionPaymentInfo.invoiceId} " +
+                        "for renewal of subscription to plan ${plan.id}"
+                )
                 plan
             }.fix()
```

Comment:
This boilerplate adds two levels of indentation and obscures control flow.  Can it be abstracted away into something more compact?